File: preopIA.js
// preopIA.js — ESM
// Ruta IA Pre Op: genera { examenes, informeIA } con OpenAI y persiste en memoria (namespace 'preop')

import OpenAI from "openai";

export default function iaPreopHandler(memoria) {
  const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
  const ns = (s, id) => `${s}:${id}`;

  // ===== Catálogo (nombres canónicos EXACTOS para compatibilidad con PDF) =====
  // Nota: incluimos entradas separadas y compuestas para mayor compatibilidad.
  const CATALOGO_EXAMENES = [
    "HEMOGRAMA MAS VHS",
    "VHS",
    "PCR",
    "GLICEMIA",
    "HEMOGLOBINA GLICOSILADA",
    "ELECTROLITOS PLASMATICOS",
    "PERFIL BIOQUIMICO",
    "PERFIL LIPIDICO",
    "PERFIL HEPATICO",
    "CREATININA",
    "UREA",
    "TP/INR",
    "TTPA",
    "TTPK",
    "PERFIL DE COAGULACION (TP/INR y TTPA)",
    "GRUPO Y RH",
    "PRUEBAS CRUZADAS (2U)",
    "VIH",
    "ORINA COMPLETA",
    "ORINA",
    "UROCULTIVO",
    "ECG DE REPOSO",
    "RADIOGRAFIA DE TORAX",
    "PASE ODONTOLOGICO",
  ];

  const CANON = new Set(CATALOGO_EXAMENES);

  // Sinónimos -> nombre canónico del catálogo
  const ALIAS = new Map([
    // Hemograma/VHS
    ["HEMOGRAMA", "HEMOGRAMA MAS VHS"],
    ["VELOCIDAD DE SEDIMENTACION", "VHS"],
    ["V.S.G.", "VHS"],
    // Glucosa
    ["GLUCOSA", "GLICEMIA"],
    // Coagulación
    ["APTT", "TTPA"],
    ["A PTT", "TTPA"],
    ["A-PTT", "TTPA"],
    ["TIEMPO DE PROTROMBINA", "TP/INR"],
    ["INR", "TP/INR"],
    ["COAGULOGRAMA", "PERFIL DE COAGULACION (TP/INR y TTPA)"],
    // Orina
    ["ORINA", "ORINA COMPLETA"],
    ["EXAMEN DE ORINA", "ORINA COMPLETA"],
    // ECG
    ["ECG", "ECG DE REPOSO"],
    ["ELECTROCARDIOGRAMA", "ECG DE REPOSO"],
    // Rx tórax
    ["RX DE TORAX", "RADIOGRAFIA DE TORAX"],
    ["RX TORAX", "RADIOGRAFIA DE TORAX"],
    ["RADIOGRAFIA TORAX", "RADIOGRAFIA DE TORAX"],
    // Odontología
    ["PASE ODONTOLÓGICO", "PASE ODONTOLOGICO"],
    ["PASE DE ODONTOLOGIA", "PASE ODONTOLOGICO"],
    ["EVALUACION ODONTOLOGICA PREOPERATORIA", "PASE ODONTOLOGICO"],
  ]);

  function normalizarNombre(raw = "") {
    const key = String(raw || "").trim().toUpperCase();
    if (CANON.has(key)) return key;          // respeta EXACTO si está en catálogo
    if (ALIAS.has(key)) return ALIAS.get(key);
    return null;
  }

  function validarContraCatalogo(lista) {
    if (!Array.isArray(lista)) return null;
    const out = new Set();
    for (const it of lista) {
      const raw = typeof it === "string" ? it : (it && it.nombre) || "";
      const norm = normalizarNombre(raw);
      if (norm) out.add(norm);
    }
    return out.size ? Array.from(out) : null;
  }

  // ===== Utilidades para informe fallback =====
  const etiqueta = {
    hta: "Hipertensión arterial",
    dm2: "Diabetes mellitus tipo 2",
    dislipidemia: "Dislipidemia",
    obesidad: "Obesidad",
    tabaquismo: "Tabaquismo",
    epoc_asma: "EPOC / Asma",
    cardiopatia: "Cardiopatía",
    erc: "Enfermedad renal crónica",
    hipotiroidismo: "Hipotiroidismo",
    anticoagulantes: "Uso de anticoagulantes/antiagregantes",
    artritis_reumatoide: "Artritis reumatoide / autoinmune",
  };

  const resumenComorbilidades = (c = {}) => {
    const pos = Object.keys(etiqueta)
      .filter((k) => c[k] === true)
      .map((k) => `• ${etiqueta[k]}`);
    return pos.length
      ? pos.join("\n")
      : "Sin comorbilidades relevantes reportadas.";
  };

  function construirInformeFallback({
    paciente = {},
    comorbilidades = {},
    tipoCirugia = "",
  }) {
    const nombre = paciente?.nombre || "";
    const edad = Number(paciente?.edad) || null;
    const dolor = paciente?.dolor || "";
    const lado = paciente?.lado || "";
    const cirugiaTxt = tipoCirugia || "No especificada";

    // Alergias: acepta string o objeto {tiene, detalle}
    const alergias =
      typeof comorbilidades?.alergias === "object"
        ? comorbilidades.alergias.tiene
          ? comorbilidades.alergias.detalle || "Refiere alergias."
          : "No refiere."
        : (comorbilidades?.alergias || "").toString().trim();

    // Anticoagulantes: acepta boolean o {usa, detalle}
    const anticoags =
      typeof comorbilidades?.anticoagulantes === "object"
        ? comorbilidades.anticoagulantes.usa
          ? `Sí${
              comorbilidades.anticoagulantes.detalle
                ? ` — ${comorbilidades.anticoagulantes.detalle}`
                : ""
            }`
          : "No"
        : comorbilidades?.anticoagulantes
        ? "Sí"
        : "No";

    const otras = (comorbilidades?.otras || "").toString().trim();
    const lista = resumenComorbilidades(comorbilidades);

    const consideraciones = [];
    if (comorbilidades.dm2) consideraciones.push("Control glicémico (HbA1c).");
    if (comorbilidades.erc)
      consideraciones.push("Evaluar función renal / evitar nefrotóxicos.");
    if (comorbilidades.cardiopatia || (edad && edad >= 60))
      consideraciones.push(
        "ECG de reposo y evaluación cardiovascular según riesgo."
      );
    if (comorbilidades.epoc_asma || comorbilidades.tabaquismo)
      consideraciones.push("Optimización respiratoria.");
    if (
      comorbilidades.anticoagulantes === true ||
      comorbilidades?.anticoagulantes?.usa
    )
      consideraciones.push("Plan suspensión o puente de anticoagulación.");

    return [
      `Evaluación Preoperatoria (resumen)\n`,
      `Paciente: ${nombre || "—"}   Edad: ${edad ?? "—"} años`,
      `Motivo/Área: ${dolor || "—"} ${lado || ""}`.trim(),
      `Cirugía planificada: ${cirugiaTxt}`,
      ``,
      `Comorbilidades:\n${lista}`,
      `Alergias: ${alergias || "—"}`,
      `Anticoagulantes/antiagregantes: ${anticoags}`,
      otras ? `Otras comorbilidades:\n${otras}` : "",
      ``,
      `Consideraciones preoperatorias:`,
      `${
        consideraciones.length
          ? "• " + consideraciones.join("\n• ")
          : "• Sin consideraciones adicionales más allá del protocolo estándar."
      }`,
    ]
      .filter(Boolean)
      .join("\n");
  }

  // ===== Reglas BASALES + por comorbilidad (deterministas) =====
  function examenesBasales(paciente = {}, comorbilidades = {}, tipoCirugia = "") {
    const edad = Number(paciente?.edad);
    const mayor60 = Number.isFinite(edad) && edad >= 60;
    const esArtro = /ARTROPLASTIA/i.test(String(tipoCirugia || ""));

    const add = new Set();

    // Basal general para cirugía mayor ortopédica (artroplastia) y preop
    add.add("HEMOGRAMA MAS VHS");
    add.add("PCR");
    add.add("GLICEMIA");
    add.add("ELECTROLITOS PLASMATICOS");
    add.add("PERFIL BIOQUIMICO");
    add.add("PERFIL HEPATICO");
    add.add("CREATININA");
    add.add("UREA");
    add.add("ORINA COMPLETA");
    add.add("UROCULTIVO");
    add.add("GRUPO Y RH");

    // *** OBLIGATORIOS: COAGULACIÓN (tres variantes para compatibilidad) ***
    add.add("PERFIL DE COAGULACION (TP/INR y TTPA)");
    add.add("TP/INR");
    add.add("TTPK");

    // *** OBLIGATORIO: ECG SIEMPRE ***
    add.add("ECG DE REPOSO");

    // Cirugía mayor: reservar cruzadas
    if (esArtro) add.add("PRUEBAS CRUZADAS (2U)");

    // Diabetes/obesidad: HbA1c
    if (comorbilidades?.dm2 || comorbilidades?.obesidad) {
      add.add("HEMOGLOBINA GLICOSILADA");
    }

    // Rx tórax si mayor60 o EPOC/asma/tabaquismo/cardiopatía
    if (
      mayor60 ||
      comorbilidades?.epoc_asma ||
      comorbilidades?.tabaquismo ||
      comorbilidades?.cardiopatia
    ) {
      add.add("RADIOGRAFIA DE TORAX");
    }

    // Pase odontológico para artroplastia (foco infeccioso)
    if (esArtro) add.add("PASE ODONTOLOGICO");

    // Validamos contra catálogo por seguridad
    return validarContraCatalogo(Array.from(add)) || [];
  }

  // ===== Prompt para ChatGPT (devuelve SOLO JSON) =====
  function buildPromptJSON({
    paciente,
    comorbilidades,
    tipoCirugia,
    catalogo,
    basales,
  }) {
    const resumen = resumenComorbilidades(comorbilidades);

    const esArtro = /ARTROPLASTIA/i.test(tipoCirugia || "");
    const reglasArtro = esArtro
      ? `
- Contexto: ARTROPLASTIA (cadera/rodilla). Los exámenes basales YA INCLUYEN (obligatorios): hemograma+VHS, PCR, glicemia, bioquímica/renal, perfil hepático, COAGULACIÓN (perfil + TP/INR + TTPK), orina completa + urocultivo, grupo y RH, ECG de reposo, pruebas cruzadas (2U), y Rx de tórax según riesgo/edad, además de HbA1c si DM/obesidad y pase odontológico.
- Tu tarea es AGREGAR del catálogo solo lo que creas faltante según comorbilidades/edad. Si nada falta, devuelve lista vacía para "examenes" (porque ya vienen los basales).`
      : `
- Hay exámenes basales predefinidos (incluyen ECG y coagulación). Solo agrega del catálogo lo que falte por comorbilidades/edad. Si nada falta, devuelve lista vacía para "examenes".`;

    return `
Eres un asistente clínico para evaluación PREOPERATORIA.
Devuelve EXCLUSIVAMENTE un JSON válido con forma:
{
  "examenes": [ /* SOLO nombres exactamente del catálogo (nuevos a agregar, los basales ya se incluyen) */ ],
  "informeIA": "texto breve en español (≤140 palabras) con consideraciones preoperatorias"
}

Reglas:
- Usa únicamente ítems del catálogo (coincidencia literal).
- No prescribas fármacos. Enfócate en laboratorio, ECG y Rx tórax si corresponde.
${reglasArtro}

Catálogo permitido:
${catalogo.map((s) => `- ${s}`).join("\n")}

Exámenes basales ya incluidos:
${basales.map((s) => `- ${s}`).join("\n")}

Datos del paciente:
- Nombre/Edad: ${paciente?.nombre || "—"} / ${Number(paciente?.edad) || "—"} años
- Cirugía planificada: ${tipoCirugia || "No especificada"}
- Motivo/Área: ${paciente?.dolor || "—"} ${paciente?.lado || ""}
- Comorbilidades marcadas:
${resumen}
`.trim();
  }

  // Robustez: extraer JSON aunque venga dentro de ```json ... ```
  function extraerJSON(str = "") {
    try {
      return JSON.parse(str);
    } catch {}
    const m =
      str.match(/```json\s*([\s\S]*?)```/i) || str.match(/```([\s\S]*?)```/);
    if (m && m[1]) {
      try {
        return JSON.parse(m[1]);
      } catch {}
    }
    const m2 = str.match(/\{[\s\S]*\}/);
    if (m2) {
      try {
        return JSON.parse(m2[0]);
      } catch {}
    }
    return null;
  }

  return async (req, res) => {
    try {
      const {
        idPago,
        paciente = {},
        comorbilidades = {},
        tipoCirugia = "",
        catalogoExamenes = [],
      } = req.body || {};

      if (!idPago || !paciente || !paciente.nombre) {
        return res
          .status(400)
          .json({ ok: false, error: "Faltan idPago o datos del paciente." });
      }

      // 1) Catálogo desde el front (si viene) o el local
      const catalogo =
        Array.isArray(catalogoExamenes) && catalogoExamenes.length
          ? catalogoExamenes.map((s) => String(s).trim()).filter(Boolean)
          : CATALOGO_EXAMENES;

      // 2) Construir basales deterministas (con ECG y coagulación obligatorios)
      const base = examenesBasales(paciente, comorbilidades, tipoCirugia);

      // 3) Llamada a OpenAI (si hay API key) para sugerir EXTRAS sobre los basales
      let extras = [];
      let informeIA = "";

      if (process.env.OPENAI_API_KEY) {
        const prompt = buildPromptJSON({
          paciente,
          comorbilidades,
          tipoCirugia,
          catalogo,
          basales: base,
        });
        try {
          const completion = await openai.chat.completions.create({
            model: "gpt-4o-mini",
            temperature: 0.2,
            max_tokens: 450,
            messages: [
              {
                role: "system",
                content:
                  "Devuelve únicamente JSON válido según las reglas dadas.",
              },
              { role: "user", content: prompt },
            ],
          });

          const content = completion?.choices?.[0]?.message?.content || "";
          const parsed = extraerJSON(content);

          if (parsed && Array.isArray(parsed.examenes)) {
            extras = validarContraCatalogo(parsed.examenes) || [];
          }
          if (parsed && typeof parsed.informeIA === "string") {
            informeIA = parsed.informeIA.trim();
          }
        } catch (err) {
          console.warn("OpenAI fallo ia-preop:", err?.message || err);
        }
      }

      // 4) Mezcla final: BASALES ∪ EXTRAS (deduplicado)
      const examenes = Array.from(new Set([...(base || []), ...(extras || [])]));

      // 5) Informe fallback si la IA no lo entregó
      if (!informeIA) {
        informeIA = construirInformeFallback({
          paciente,
          comorbilidades,
          tipoCirugia,
        });
      }

      // 6) Persistencia (merge sin destruir lo previo)
      const prev = memoria.get(ns("preop", idPago)) || {};
      const next = {
        ...prev,
        ...paciente, // aplanado
        comorbilidades,
        tipoCirugia,
        examenesIA: examenes,
        informeIA,
      };
      memoria.set(ns("preop", idPago), next);

      return res.json({ ok: true, examenes, informeIA });
    } catch (e) {
      console.error("ia-preop error:", e);
      return res.status(500).json({
        ok: false,
        error: "No se pudo generar indicación preoperatoria.",
      });
    }
  };
}

File: ia/preop_ia.py
# ia/preop_ia.py
# IA Pre Op: recibe { idPago, paciente, comorbilidades, tipoCirugia }
# Devuelve { ok, examenes, informeIA }
# Stateless — sin memoria server-side

import re
import json
import logging
from typing import Any

import httpx

logger = logging.getLogger("preop_ia")

# ============================================================
# CATÁLOGO EXACTO
# ============================================================
CATALOGO_EXAMENES: list[str] = [
    "HEMOGRAMA MAS VHS",
    "VHS",
    "PCR",
    "GLICEMIA",
    "HEMOGLOBINA GLICOSILADA",
    "ELECTROLITOS PLASMATICOS",
    "PERFIL BIOQUIMICO",
    "PERFIL LIPIDICO",
    "PERFIL HEPATICO",
    "CREATININA",
    "UREA",
    "TP/INR",
    "TTPA",
    "TTPK",
    "PERFIL DE COAGULACION (TP/INR y TTPA)",
    "GRUPO Y RH",
    "PRUEBAS CRUZADAS (2U)",
    "VIH",
    "ORINA COMPLETA",
    "ORINA",
    "UROCULTIVO",
    "ECG DE REPOSO",
    "RADIOGRAFIA DE TORAX",
    "PASE ODONTOLOGICO",
]

_CANON: set[str] = set(CATALOGO_EXAMENES)

_ALIAS: dict[str, str] = {
    "HEMOGRAMA":                              "HEMOGRAMA MAS VHS",
    "VELOCIDAD DE SEDIMENTACION":             "VHS",
    "V.S.G.":                                 "VHS",
    "GLUCOSA":                                "GLICEMIA",
    "APTT":                                   "TTPA",
    "A PTT":                                  "TTPA",
    "A-PTT":                                  "TTPA",
    "TIEMPO DE PROTROMBINA":                  "TP/INR",
    "INR":                                    "TP/INR",
    "COAGULOGRAMA":                           "PERFIL DE COAGULACION (TP/INR y TTPA)",
    "ORINA":                                  "ORINA COMPLETA",
    "EXAMEN DE ORINA":                        "ORINA COMPLETA",
    "ECG":                                    "ECG DE REPOSO",
    "ELECTROCARDIOGRAMA":                     "ECG DE REPOSO",
    "RX DE TORAX":                            "RADIOGRAFIA DE TORAX",
    "RX TORAX":                               "RADIOGRAFIA DE TORAX",
    "RADIOGRAFIA TORAX":                      "RADIOGRAFIA DE TORAX",
    "PASE ODONTOLÓGICO":                      "PASE ODONTOLOGICO",
    "PASE DE ODONTOLOGIA":                    "PASE ODONTOLOGICO",
    "EVALUACION ODONTOLOGICA PREOPERATORIA":  "PASE ODONTOLOGICO",
}

_ETIQUETAS: dict[str, str] = {
    "hta":                 "Hipertensión arterial",
    "dm2":                 "Diabetes mellitus tipo 2",
    "dislipidemia":        "Dislipidemia",
    "obesidad":            "Obesidad",
    "tabaquismo":          "Tabaquismo",
    "epoc_asma":           "EPOC / Asma",
    "cardiopatia":         "Cardiopatía",
    "erc":                 "Enfermedad renal crónica",
    "hipotiroidismo":      "Hipotiroidismo",
    "anticoagulantes":     "Uso de anticoagulantes/antiagregantes",
    "artritis_reumatoide": "Artritis reumatoide / autoinmune",
}


# ============================================================
# HELPERS
# ============================================================
def _normalizar_nombre(raw: str) -> str | None:
    key = str(raw or "").strip().upper()
    if key in _CANON:
        return key
    return _ALIAS.get(key)


def _validar_contra_catalogo(lista: list[Any]) -> list[str]:
    out: list[str] = []
    seen: set[str] = set()
    for it in lista:
        raw = it if isinstance(it, str) else (it or {}).get("nombre", "")
        norm = _normalizar_nombre(str(raw))
        if norm and norm not in seen:
            out.append(norm)
            seen.add(norm)
    return out


def _resumen_comorbilidades(c: dict) -> str:
    pos = [f"• {_ETIQUETAS[k]}" for k in _ETIQUETAS if c.get(k) is True]
    return "\n".join(pos) if pos else "Sin comorbilidades relevantes reportadas."


def _texto_alergias(c: dict) -> str:
    al = c.get("alergias", "")
    if isinstance(al, dict):
        return al.get("detalle") or "Refiere alergias." if al.get("tiene") else "No refiere."
    return str(al).strip() or "—"


def _texto_anticoagulantes(c: dict) -> str:
    ac = c.get("anticoagulantes", "")
    if isinstance(ac, dict):
        if ac.get("usa"):
            detalle = ac.get("detalle", "")
            return f"Sí — {detalle}" if detalle else "Sí"
        return "No"
    return "Sí" if ac else "No"


def _safe_edad(paciente: dict) -> int | None:
    try:
        return int(paciente.get("edad") or 0) or None
    except (ValueError, TypeError):
        return None


def _es_artro(tipo_cirugia: str) -> bool:
    return bool(re.search(r"ARTROPLASTIA", str(tipo_cirugia or "").upper()))


# ============================================================
# INFORME FALLBACK
# ============================================================
def _construir_informe_fallback(paciente: dict, comorbilidades: dict, tipo_cirugia: str) -> str:
    nombre      = paciente.get("nombre") or ""
    edad        = _safe_edad(paciente)
    dolor       = paciente.get("dolor") or "—"
    lado        = paciente.get("lado") or ""
    cirugia_txt = tipo_cirugia or "No especificada"
    lista       = _resumen_comorbilidades(comorbilidades)
    alergias    = _texto_alergias(comorbilidades)
    anticoags   = _texto_anticoagulantes(comorbilidades)
    otras       = str(comorbilidades.get("otras") or "").strip()

    c = comorbilidades
    consideraciones = []
    if c.get("dm2"):
        consideraciones.append("Control glicémico (HbA1c).")
    if c.get("erc"):
        consideraciones.append("Evaluar función renal / evitar nefrotóxicos.")
    if c.get("cardiopatia") or (edad and edad >= 60):
        consideraciones.append("ECG de reposo y evaluación cardiovascular según riesgo.")
    if c.get("epoc_asma") or c.get("tabaquismo"):
        consideraciones.append("Optimización respiratoria.")
    if c.get("anticoagulantes") is True or (isinstance(c.get("anticoagulantes"), dict) and c["anticoagulantes"].get("usa")):
        consideraciones.append("Plan suspensión o puente de anticoagulación.")

    lineas = [
        "Evaluación Preoperatoria (resumen)\n",
        f"Paciente: {nombre or '—'}   Edad: {edad or '—'} años",
        f"Motivo/Área: {dolor} {lado}".strip(),
        f"Cirugía planificada: {cirugia_txt}",
        "",
        f"Comorbilidades:\n{lista}",
        f"Alergias: {alergias}",
        f"Anticoagulantes/antiagregantes: {anticoags}",
    ]
    if otras:
        lineas.append(f"Otras comorbilidades:\n{otras}")
    lineas += [
        "",
        "Consideraciones preoperatorias:",
        ("• " + "\n• ".join(consideraciones)) if consideraciones
        else "• Sin consideraciones adicionales más allá del protocolo estándar.",
    ]
    return "\n".join(lineas)


# ============================================================
# BASALES DETERMINISTAS
# ============================================================
def _examenes_basales(paciente: dict, comorbilidades: dict, tipo_cirugia: str) -> list[str]:
    edad    = _safe_edad(paciente)
    mayor60 = edad is not None and edad >= 60
    artro   = _es_artro(tipo_cirugia)
    c       = comorbilidades

    add: set[str] = {
        "HEMOGRAMA MAS VHS", "PCR", "GLICEMIA",
        "ELECTROLITOS PLASMATICOS", "PERFIL BIOQUIMICO", "PERFIL HEPATICO",
        "CREATININA", "UREA", "ORINA COMPLETA", "UROCULTIVO", "GRUPO Y RH",
        "PERFIL DE COAGULACION (TP/INR y TTPA)", "TP/INR", "TTPK",
        "ECG DE REPOSO",
    }

    if artro:
        add.add("PRUEBAS CRUZADAS (2U)")
        add.add("PASE ODONTOLOGICO")

    if c.get("dm2") or c.get("obesidad"):
        add.add("HEMOGLOBINA GLICOSILADA")

    if mayor60 or c.get("epoc_asma") or c.get("tabaquismo") or c.get("cardiopatia"):
        add.add("RADIOGRAFIA DE TORAX")

    return _validar_contra_catalogo(list(add))


# ============================================================
# PROMPT IA
# ============================================================
def _build_prompt(paciente: dict, comorbilidades: dict, tipo_cirugia: str,
                  catalogo: list[str], basales: list[str]) -> str:
    resumen  = _resumen_comorbilidades(comorbilidades)
    artro    = _es_artro(tipo_cirugia)
    edad     = _safe_edad(paciente) or "—"
    nombre   = paciente.get("nombre") or "—"
    dolor    = paciente.get("dolor") or "—"
    lado     = paciente.get("lado") or ""
    cat_str  = "\n".join(f"- {s}" for s in catalogo)
    base_str = "\n".join(f"- {s}" for s in basales)

    reglas_artro = (
        "- Contexto: ARTROPLASTIA. Los basales ya incluyen coagulación, cruzadas, odontológico. "
        "Solo agrega lo faltante por comorbilidades. Si nada falta, devuelve lista vacía."
        if artro else
        "- Hay basales predefinidos. Solo agrega del catálogo lo faltante por comorbilidades/edad. "
        "Si nada falta, devuelve lista vacía."
    )

    return f"""Eres un asistente clínico para evaluación PREOPERATORIA.
Devuelve EXCLUSIVAMENTE un JSON válido:
{{
  "examenes": [ /* SOLO nombres exactos del catálogo, adicionales a los basales */ ],
  "informeIA": "texto breve en español (≤140 palabras) con consideraciones preoperatorias"
}}

Reglas:
- Usa únicamente ítems del catálogo (coincidencia literal).
- No prescribas fármacos.
{reglas_artro}

Catálogo permitido:
{cat_str}

Exámenes basales ya incluidos:
{base_str}

Datos:
- Paciente: {nombre} / {edad} años
- Cirugía planificada: {tipo_cirugia or "No especificada"}
- Motivo/Área: {dolor} {lado}
- Comorbilidades:
{resumen}""".strip()


# ============================================================
# EXTRACCIÓN ROBUSTA DE JSON
# ============================================================
def _extraer_json(texto: str) -> dict | None:
    for candidato in [
        texto,
        (re.search(r"```json\s*([\s\S]*?)```", texto, re.I) or re.search(r"```([\s\S]*?)```", texto) or type("", (), {"group": lambda s, n: None})()).group(1),
        (re.search(r"\{[\s\S]*\}", texto) or type("", (), {"group": lambda s, n: None})()).group(0),
    ]:
        if not candidato:
            continue
        try:
            return json.loads(candidato)
        except (json.JSONDecodeError, TypeError):
            continue
    return None


# ============================================================
# LLAMADAS IA
# ============================================================
async def _llamar_claude(prompt: str, api_key: str, model: str) -> str:
    async with httpx.AsyncClient(timeout=30.0) as client:
        r = await client.post(
            "https://api.anthropic.com/v1/messages",
            headers={
                "x-api-key":         api_key,
                "anthropic-version": "2023-06-01",
                "content-type":      "application/json",
            },
            json={
                "model":      model,
                "max_tokens": 500,
                "system":     "Devuelve únicamente JSON válido según las reglas dadas.",
                "messages":   [{"role": "user", "content": prompt}],
            },
        )
        r.raise_for_status()
        return (r.json().get("content") or [{}])[0].get("text", "").strip()


async def _llamar_openai(prompt: str, api_key: str, model: str) -> str:
    async with httpx.AsyncClient(timeout=30.0) as client:
        r = await client.post(
            "https://api.openai.com/v1/chat/completions",
            headers={
                "authorization": f"Bearer {api_key}",
                "content-type":  "application/json",
            },
            json={
                "model":       model,
                "temperature": 0.2,
                "max_tokens":  450,
                "messages": [
                    {"role": "system", "content": "Devuelve únicamente JSON válido según las reglas dadas."},
                    {"role": "user",   "content": prompt},
                ],
            },
        )
        r.raise_for_status()
        return (r.json().get("choices") or [{}])[0].get("message", {}).get("content", "").strip()


# ============================================================
# HANDLER PRINCIPAL
# ============================================================
async def preop_ia(payload: dict, config: dict) -> dict:
    """
    payload: { idPago, paciente, comorbilidades, tipoCirugia?, catalogoExamenes? }
    config:  { anthropic_api_key?, openai_api_key?, anthropic_model?, openai_model? }
    Retorna: { ok, examenes, informeIA }
    """
    id_pago        = payload.get("idPago") or ""
    paciente       = payload.get("paciente") or {}
    comorbilidades = payload.get("comorbilidades") or {}
    tipo_cirugia   = str(payload.get("tipoCirugia") or "")
    catalogo_raw   = payload.get("catalogoExamenes") or []

    if not id_pago or not paciente.get("nombre"):
        return {"ok": False, "error": "Faltan idPago o datos del paciente."}

    catalogo = (
        [str(s).strip() for s in catalogo_raw if s]
        if catalogo_raw else CATALOGO_EXAMENES
    )

    anthropic_key   = config.get("anthropic_api_key") or ""
    openai_key      = config.get("openai_api_key") or ""
    anthropic_model = config.get("anthropic_model") or "claude-sonnet-4-6"
    openai_model    = config.get("openai_model") or "gpt-4o-mini"

    # 1) Basales deterministas
    base = _examenes_basales(paciente, comorbilidades, tipo_cirugia)

    # 2) IA para extras sobre los basales
    extras: list[str] = []
    informe_ia = ""
    prompt = _build_prompt(paciente, comorbilidades, tipo_cirugia, catalogo, base)

    if anthropic_key:
        try:
            texto  = await _llamar_claude(prompt, anthropic_key, anthropic_model)
            parsed = _extraer_json(texto)
            if parsed and isinstance(parsed.get("examenes"), list):
                extras = _validar_contra_catalogo(parsed["examenes"])
            if parsed and isinstance(parsed.get("informeIA"), str):
                informe_ia = parsed["informeIA"].strip()
        except Exception as e:
            logger.warning("Claude falló ia-preop: %s", e)

    if not extras and openai_key:
        try:
            texto  = await _llamar_openai(prompt, openai_key, openai_model)
            parsed = _extraer_json(texto)
            if parsed and isinstance(parsed.get("examenes"), list):
                extras = _validar_contra_catalogo(parsed["examenes"])
            if parsed and isinstance(parsed.get("informeIA"), str):
                informe_ia = parsed["informeIA"].strip()
        except Exception as e:
            logger.warning("OpenAI falló ia-preop: %s", e)

    # 3) Mezcla final: basales ∪ extras (deduplicado, orden preservado)
    seen: set[str] = set()
    examenes: list[str] = []
    for e in base + extras:
        if e not in seen:
            examenes.append(e)
            seen.add(e)

    # 4) Informe fallback si IA no entregó
    if not informe_ia:
        informe_ia = _construir_informe_fallback(paciente, comorbilidades, tipo_cirugia)

    return {"ok": True, "examenes": examenes, "informeIA": informe_ia}
  
